Initial benchmarking analysis

d33bs opened the pull request:
This PR introduces benchmarking analysis in the form of Jupyter notebooks and related charts which demonstrate various technical design capabilities found within CytoTable. Here I focused on the core technologies used by CytoTable to help demonstrate core aspects of "data grammar" implemented using Apache Arrow, DuckDB SQL, and Apache Parquet.

Thank you for any feedback you may have on this! Please don't hesitate to let me know if you have any questions.

d33bs committed:
initial analysis work

d33bs committed:
linting

d33bs committed:
arrow analysis; restructure for utils

d33bs committed:
linting

d33bs committed:
chart adjustments and images exports

d33bs committed:
updated colors

d33bs committed:
chart updates

d33bs committed:
data prep

d33bs committed:
add dots

d33bs committed:
update images; add duckdb analysis

d33bs committed:
linting

gwaybio commented on notebooks/arrow_analysis.py:
```diff
@@ -0,0 +1,143 @@
+        size=17.5,  # global font size
+    ),
+)
+fig.show()
```
These aren't rendering in the notebook, which would be helpful to see. Consider adding in the notebook as well.

d33bs replied:
Thanks! I've made some updates to include the saved images instead, which I think might be a better option than the dynamic ones for the usecase.

gwaybio commented on notebooks/arrow_analysis.py:
```diff
@@ -0,0 +1,143 @@
+)
+fig.show()
+
+pio.write_image(fig, f"{image_dir}/mem-read-times.png")
```
FWIW, I tend to prefer defining all input/output files in an early cell (after importing) so that folks can easily see what i/o to expect. Consider doing that here, although totally optional

d33bs replied:
Thanks, this makes a lot of sense, especially to avoid repetition and help with expectation-setting upfront, as you mentioned.

gwaybio commented on notebooks/create_example_data.py:
```diff
@@ -0,0 +1,135 @@
+orig_filepath = "./examples/data/all_cellprofiler.sqlite"
+
+
+def download_file(url, filename):
```
you probably don't need to make the change in this PR (in case you never need to run the function again), but it might be worth importing this function from a `utils.py` file, for example.

d33bs replied:
Great thought! I've added this to the utils.

gwaybio commented on notebooks/create_example_data.py:
```diff
@@ -0,0 +1,135 @@
+download_file(url, orig_filepath)
+
+
+def double_database_size(filename: str) -> str:
```
same goes for this function.

Also, is it worth renaming to `multiply_database_size()`, and then have an argument to specify if double, triple, etc.?

Definitely an optional comment, btw. (You will need to decide if this is worth implementing, and I'd say this is low priority)

d33bs replied:
Cheers, love this idea, will modify the function as such. I ended up also refactoring to use a for loop with the doubling efforts.

gwaybio commented on notebooks/create_example_data.py:
```diff
@@ -0,0 +1,135 @@
+double_database_size(filename=doubled_file)
+
+# # copy the original with new name
+quadrupled_file = shutil.copy(
```
ha! Or you could do it like how you do here! :) (in reference to previous comment)

d33bs replied:
🙂

gwaybio commented on notebooks/examples/join_duckdb.py:
```diff
@@ -0,0 +1,32 @@
+import sys
```
Love having the example files here! Consider adding a docstring describing the purpose to make this explicit

d33bs replied:
Great point, thanks!

gwaybio commented on notebooks/examples/join_pandas.py:
```diff
@@ -0,0 +1,34 @@
+import sys
```
same docstring comment here

gwaybio commented on notebooks/examples/join_pandas.py:
```diff
@@ -0,0 +1,34 @@
+import sys
+
+import pandas as pd
```
I'm also wondering if it's worth including a `pycytominer` merge here 🤔 (maybe not worth it, but maybe something to consider)

d33bs replied:
Thanks I think this is a good idea! I tried to focus on the most basic examples in this PR, building a foundation of benchmarks on underlying technologies used, and would like to follow up with more complex analysis of `pycytominer` and others. I'll open an issue to explore this further.

gwaybio reviewed (approved):
This is beautiful work! Very well documented, and a joy to read.

I made a couple inline comments, but none that need to delay merging.

Here are also a couple other, general comments:

- I don't see the `.py` file for [notebooks/duckdb_analysis.ipynb](https://github.com/cytomining/CytoTable-benchmarks/pull/1/files#diff-f19aa6b517dbec6fed032a2cb2b4b76fba1392c3eda1227381d3ac560fd920d9)
- It's great to see the different sizes change performance, and how your code can handle the differing sizes really elegantly. I think it is worth increasing the largest again, by at least double. We regularly see CP files with hundreds of thousands of rows (feel free to check with @jenna-tomkinson regarding some more real-world dataset expectations), and it would be great to show some on the larger end. (if the concern is only displaying a focused set of three comparisons, perhaps remove the low in favor of a higher size; although I think 4 or five would be fine 😄 )
- It might help to clarify some of the figures. It wasn't clear to me immediately if the figures were describing `all` or `one` sometimes. Also, it was unclear to me why sometimes we were comparing `sqlite`, `csv` and `parquet` and other times `pandas`, `polars` and `pyarrow`. Maybe being more specific about what you mean by `variable` in the plot legend would help?

d33bs committed:
update images; utils; sqlite data generation

d33bs committed:
modify layout

d33bs committed:
database multiplier instead of static doubler

Co-Authored-By: Gregory Way <gregory.way@gmail.com>

d33bs committed:
addressing pr comments

d33bs committed:
new chart names

d33bs commented:
Thanks @gwaybio ! I've addressed many of the comments and added one issue for future work surrounding more complex analysis I hope we can achieve.

> I don't see the .py file for notebooks/duckdb_analysis.ipynb

Thanks! I've added this into the changes.

> It's great to see the different sizes change performance, and how your code can handle the differing sizes really elegantly. I think it is worth increasing the largest again, by at least double. We regularly see CP files with hundreds of thousands of rows (feel free to check with @jenna-tomkinson regarding some more real-world dataset expectations), and it would be great to show some on the larger end. (if the concern is only displaying a focused set of three comparisons, perhaps remove the low in favor of a higher size; although I think 4 or five would be fine 😄 )

Great thought! I've added an "x32" version of the database file used for those benchmarks, further exhibiting the patterns. During development I had intended to observe both column and row multiplication exponentially but this proved challenging due to [SQLite column number default maximum (2000)](https://www.sqlite.org/limits.html#max_column) and my attempts here to simplify or avoid design choice invalidation (I wasn't sure if these defaults were somehow related to performance or file size in the SQLite codebase). That said, I don't see a reason we couldn't increase the row count further, but this may impact testing time durations more generally.

> It might help to clarify some of the figures. It wasn't clear to me immediately if the figures were describing all or one sometimes. Also, it was unclear to me why sometimes we were comparing sqlite, csv and parquet and other times pandas, polars and pyarrow. Maybe being more specific about what you mean by variable in the plot legend would help?

Cheers, this is great feedback! I've updated all the chart names and legend titles for greater understandability.
